fix(ivy): do not reset view dirty state in check no changes mode

devversion opened the pull request:
In Ivy we currently also reset the dirty state of components in the check
no changes change detection cycle. This is not the case in View Engine.

This behavior in Ivy results in inconsistencies and unpredictable behavior of
components using the `OnPush` change detection strategy. One example
is that components can no longer be marked as dirty from view lifecycle hooks
because the dirty flag is reset afterwards in the check-no-changes CD cycle.

Resolves FW-1354. Fixes #34471

devversion commented on packages/core/test/acceptance/integration_spec.ts:
```diff
@@ -1973,4 +1972,103 @@ describe('acceptance integration tests', () => {
+
+            constructor(private _cdRef: ChangeDetectorRef){}
+
+                [hookName]() {
```
FWIW: The weird formatting comes from clang-format.

devversion commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -475,7 +475,15 @@ export function refreshView<T>(
+    // result in inconsistent and unpredictable behavior when check no changes is disabled.
+    // View Engine did not reset the dirty state in check no changes mode either.
+    if (!checkNoChangesMode) {
+      lView[FLAGS] &= ~LViewFlags.Dirty;
```
I was unsure if we would want to do the same for `FirstLViewPass` flag unset, but I decided to keep that as is.

pkozlowski-opensource replied:
I would rather move both flags setting into this if - otherwise we have to read / write to the  `lView[FLAGS]` multiple times (and this is expensive).

pkozlowski-opensource commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -475,7 +475,15 @@ export function refreshView<T>(
+
+    // Do not reset the dirty state when running in check no changes mode. Resetting would
+    // result in inconsistent and unpredictable behavior when check no changes is disabled.
+    // View Engine did not reset the dirty state in check no changes mode either.
```
I would drop the "View Engine did not reset the dirty state in check no changes mode either." from the documentation comment (and maybe move it to the commit message) - as the time goes by we will delete VE and those comments won't be relevant any more.

pkozlowski-opensource commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -475,7 +475,15 @@ export function refreshView<T>(
+    lView[FLAGS] &= ~LViewFlags.FirstLViewPass;
+
+    // Do not reset the dirty state when running in check no changes mode. Resetting would
+    // result in inconsistent and unpredictable behavior when check no changes is disabled.
```
nit: when I read this comment it sounds like "don't touch or bad things will happen!". While it serves as a warning I think it would be better to express it as follows: we are doing things this way since we want the following scenario to work correctly (describe scenario here).

Can you think of re-phrasing this comment so we've got the exact scenario here?

devversion replied:
Done. good idea with providing a concrete scenario.

pkozlowski-opensource reviewed (changes requested):
@devversion the change LGTM overall but I would like to move flags resetting to one place (perf).

kara commented on packages/core/test/acceptance/integration_spec.ts:
```diff
@@ -1973,4 +1972,103 @@ describe('acceptance integration tests', () => {
+      const detectChanges = (f: ComponentFixture<any>) => f.detectChanges(checkNoChanges);
+
+      // 1. ngAfterViewInit and ngAfterViewChecked lifecycle hooks run after "OnPushComp" has
+      //    been refreshed. They can leave the component as dirty. Meaning that the "OnPushComp"
```
```suggestion
      //    been refreshed. They can mark the component as dirty. Meaning that the "OnPushComp"
```

kara commented on packages/core/test/acceptance/integration_spec.ts:
```diff
@@ -1973,4 +1972,103 @@ describe('acceptance integration tests', () => {
+      //    been refreshed. They can leave the component as dirty. Meaning that the "OnPushComp"
+      //    can be checked/refreshed in a subsequent change detection cycle.
+      // 2. ngDoCheck and ngAfterContentChecked lifecycle hooks run before "OnPushComp" is
+      //    refreshed. This means that those hooks cannot leave the component as dirty because
```
```suggestion
      //    refreshed. This means that those hooks cannot mark the component as dirty because
```

kara reviewed (approved):
LGTM once my comments and Pawel's are addressed

devversion committed:
fix(ivy): do not reset view dirty state in check no changes mode

Unlike in View Engine, we currently reset the dirty state of
components in the check no changes change detection cycle.

This means that components cannot be marked as dirty from
view lifecycle hooks because the dirty state is reset and
the lifecycle hooks do not run in the check no changes CD cycle.

devversion commented:
@kara @pkozlowski-opensource Done. Thanks for reviewing! Yeah, I was thinking about combining both flag resets for perf, but wasn't sure if it semantically has any impact. I combined them now.

AndrewKushnir commented:
[Presubmit](http://test/OCL:289271760:BASE:289271763:1578782848973:ff60298d)

pkozlowski-opensource reviewed (approved):
LGTM, thnx for addressing all the comments @devversion !
